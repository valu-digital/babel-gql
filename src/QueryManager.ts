export interface QueryManagerOptions {
    onWriteQueryFile(query: {
        query: string;
        queryName: string;
        queryId: string;
    }): Promise<any>;
}

interface QueryData {
    query: string;
    queryId: string;
    queryName: string;
    requiredFragments: string[];
}

/**
 * In memory presentation of GraphQL queries that appear in the code
 */
export class QueryManager {
    queries = new Map<string, QueryData | undefined>();

    fragments = new Map<
        string,
        | {
              fragment: string;
              usedByQueries: Set<string>;
          }
        | undefined
    >();

    options: QueryManagerOptions;

    constructor(options: QueryManagerOptions) {
        this.options = options;
    }

    parseGQLTag(quasi: string) {}

    addQuery(query: string) {}

    addFragment(fragment: string) {}

    hasRequiredFragments(queryName: string) {
        const query = this.queries.get(queryName);
        if (!query) {
            return false;
        }

        return query.requiredFragments.every(fragmentName =>
            this.fragments.has(fragmentName),
        );
    }

    formatQuery(query: QueryData) {
        const fragments = query.requiredFragments
            .map(fragmentName => this.fragments.get(fragmentName))
            .join("\n");

        return `${fragments}\n${query.query}`;
    }

    async writeQueryFile(queryName: string) {
        if (!this.hasRequiredFragments(queryName)) {
            return;
        }

        const query = this.queries.get(queryName);

        if (!query) {
            return;
        }

        await this.options.onWriteQueryFile({
            query: this.formatQuery(query),
            queryId: query.queryId,
            queryName: query.queryName,
        });
    }
}
