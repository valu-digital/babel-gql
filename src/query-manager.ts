import {
    parse,
    print,
    FragmentDefinitionNode,
    DocumentNode,
    visit,
    OperationDefinitionNode,
    isOutputType,
} from "graphql";

function findFragmentDefinitions(
    docs: (DocumentNode | FragmentDefinitionNode)[],
) {
    const fragments = new Map<string, FragmentDefinitionNode>();

    for (const doc of docs) {
        visit(doc, {
            FragmentDefinition: {
                enter(node) {
                    fragments.set(node.name.value, node);
                },
            },
        });
    }

    return fragments;
}

export function findUsedFragments(
    operation: OperationDefinitionNode | FragmentDefinitionNode,
    knownFragments: ReadonlyMap<string, FragmentDefinitionNode>,
    _usedFragments?: Map<string, FragmentDefinitionNode>,
) {
    const usedFragments = _usedFragments
        ? _usedFragments
        : new Map<string, FragmentDefinitionNode>();

    visit(operation, {
        FragmentSpread: {
            enter(node) {
                const frag = knownFragments.get(node.name.value);
                if (frag) {
                    usedFragments.set(node.name.value, frag);
                    findUsedFragments(frag, knownFragments, usedFragments);
                } else {
                    throw new Error("Unknown fragment: " + node.name.value);
                }
            },
        },
    });

    return usedFragments;
}

export interface QueryManagerOptions {
    onExportQuery(
        query: {
            query: string;
            fragments: string[];
        },
        queryManager: QueryManager,
    ): Promise<any>;
}

interface QueryData {
    query: string;
    queryId: string;
    queryName: string;
    requiredFragments: string[];
}

function isOperationDefinition(ob: any): ob is OperationDefinitionNode {
    return Boolean(ob && ob.kind === "OperationDefinition");
}

function isFragmentDefinition(ob: any): ob is FragmentDefinitionNode {
    return Boolean(ob && ob.kind === "FragmentDefinition");
}

/**
 * In memory presentation of GraphQL queries that appear in the code
 */
export class QueryManager {
    queries = new Map<string, string | undefined>();
    knownFragments = new Map<string, string | undefined>();
    fragmentsUsedByQuery = new Map<string, Set<string> | undefined>();
    fragmentsUsedByFragment = new Map<string, Set<string> | undefined>();
    dirtyQueries = new Set<string>();

    options: QueryManagerOptions;

    constructor(options: QueryManagerOptions) {
        this.options = options;
    }

    parseGraphQL(graphql: string) {
        const doc = parse(graphql);

        visit(doc, {
            OperationDefinition: def => {
                if (!def.name) {
                    throw new Error("OperationDefinition missing name");
                }

                this.dirtyQueries.add(def.name.value);
                this.queries.set(def.name.value, print(def).trim());
                this.fragmentsUsedByQuery.set(def.name.value, new Set());
            },
            FragmentDefinition: def => {
                const fragmentName = def.name.value;
                this.knownFragments.set(fragmentName, print(def).trim());
                this.fragmentsUsedByFragment.set(fragmentName, new Set());

                const relatedFragments = new Set<string>();
                relatedFragments.add(fragmentName);

                for (const [
                    relatedFragmentName,
                    usedFragmentNames,
                ] of this.fragmentsUsedByFragment.entries()) {
                    if (!usedFragmentNames) {
                        continue;
                    }
                    if (usedFragmentNames.has(fragmentName)) {
                        relatedFragments.add(relatedFragmentName);
                    }
                }

                for (const [
                    queryName,
                    usedFragmentNames,
                ] of this.fragmentsUsedByQuery.entries()) {
                    if (!usedFragmentNames) {
                        continue;
                    }

                    for (const usedFragmentName of usedFragmentNames) {
                        if (relatedFragments.has(usedFragmentName)) {
                            this.dirtyQueries.add(queryName);
                        }
                    }
                }
            },
            FragmentSpread: (node, key, parent, path, ancestors) => {
                const fragmentSpreadName = node.name.value;

                for (const a of ancestors) {
                    if (isFragmentDefinition(a)) {
                        const parentFragmentName = a.name.value;

                        let fragments = this.fragmentsUsedByFragment.get(
                            parentFragmentName,
                        );

                        if (!fragments) {
                            fragments = new Set();
                            this.fragmentsUsedByFragment.set(
                                parentFragmentName,
                                fragments,
                            );
                        }

                        fragments.add(fragmentSpreadName);
                    }

                    if (isOperationDefinition(a)) {
                        const queryName = a.name?.value;
                        if (!queryName) {
                            continue;
                        }

                        let fragments = this.fragmentsUsedByQuery.get(
                            queryName,
                        );

                        if (!fragments) {
                            fragments = new Set();
                            this.fragmentsUsedByQuery.set(queryName, fragments);
                        }

                        fragments.add(fragmentSpreadName);
                    }
                }
            },
        });
    }

    queryHasRequiredFragments(queryName: string) {
        const usedFragments = this.getUsedFragmentNamesForQuery(queryName);

        for (const fragmentName of usedFragments) {
            if (!this.knownFragments.has(fragmentName)) {
                return false;
            }
        }

        return true;
    }

    /**
     * Get set of dirty queries that have all required fragments
     */
    popDirtyQueries() {
        const availableQueries = this.dirtyQueries;
        const popQueries = new Set<string>();
        this.dirtyQueries = new Set();

        for (const queryName of availableQueries) {
            if (this.queryHasRequiredFragments(queryName)) {
                popQueries.add(queryName);
            } else {
                this.dirtyQueries.add(queryName);
            }
        }

        return popQueries;
    }

    async exportDirtyQueries() {
        for (const queryName of this.popDirtyQueries()) {
            await this.exportQuery(queryName);
        }
    }

    getUsedFragmentNamesForQuery(queryName: string) {
        const fragmentNames = new Set<string>();
        const usedFragments = this.fragmentsUsedByQuery.get(queryName);

        if (!usedFragments) {
            return fragmentNames;
        }

        for (const usedFragmentName of usedFragments) {
            fragmentNames.add(usedFragmentName);
            this.getNestedFragmentNamesForFragment(
                usedFragmentName,
                fragmentNames,
            );
        }

        return fragmentNames;
    }

    getNestedFragmentNamesForFragment(
        fragmentName: string,
        _fragments?: Set<string>,
    ) {
        const fragments = _fragments || new Set();
        const usedFragments = this.fragmentsUsedByFragment.get(fragmentName);

        if (usedFragments) {
            for (const usedFragmentName of usedFragments) {
                fragments.add(usedFragmentName);
                this.getNestedFragmentNamesForFragment(
                    usedFragmentName,
                    fragments,
                );
            }
        }

        return fragments;
    }

    async exportQuery(queryName: string) {
        // if (!this.hasRequiredFragments(queryName)) {
        //     return;
        // }

        const query = this.queries.get(queryName);

        if (!query) {
            return;
        }

        const fragments = Array.from(
            this.getUsedFragmentNamesForQuery(queryName),
        )
            .map(fragmentName => {
                return this.knownFragments.get(fragmentName)!;
            })
            .sort();

        await this.options.onExportQuery(
            {
                query,
                fragments,
                // query: this.formatQuery(query),
                // queryId: query.queryId,
                // queryName: query.queryName,
            },
            this,
        );
    }
}
