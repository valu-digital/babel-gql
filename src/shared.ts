export interface ParsedGQLTag {
    queries: {
        query: string;
        queryName: string;
        queryId: string;
        usedFragments: string[];
    }[];
    fragments: {
        fragment: string;
        fragmentName: string;
        fragmentId: string;
        usedFragments: string[];
    }[];
}

export function combinedIds(ids: string[]) {
    const [first, ...rest] = ids;

    if (rest.length === 0) {
        return first;
    }

    const chunkSize = Math.floor(first.length / ids.length);

    return (
        first.slice(0, chunkSize) +
        rest
            .sort()
            .map(id => id.slice(0, chunkSize))
            .join("")
    );
}
