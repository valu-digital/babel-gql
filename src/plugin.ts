import { promises as fs } from "fs";
import PathUtils from "path";
import * as BabelTypes from "@babel/types";
import { Visitor, NodePath } from "@babel/traverse";
import { QueryManager } from "./query-manager";

export interface Babel {
    types: typeof BabelTypes;
}

interface BabelFile {
    opts: {
        generatorOpts: any;
    };
    code: string;
    path: NodePath;
}

export interface BabelGQLOptions {
    ownModuleName?: string;
    export?: boolean;
    removeQuery?: boolean;
    target?: string;
}

interface VisitorState {
    opts?: BabelGQLOptions;
    file: BabelFile;
}

function parseTag(
    t: typeof BabelTypes,
    qm: QueryManager,
    path: NodePath<BabelTypes.TaggedTemplateExpression>,
    removeQuery: boolean,
) {
    const gqlString = path.node.quasi.quasis
        .map(q => q.value.cooked)
        .join("")
        .trim();

    let parsed;

    try {
        parsed = qm.parseGraphQL(gqlString);
    } catch (error) {
        throw path.buildCodeFrameError(
            "GraphQL: " + error.message ?? "Unknown graphql parsing error",
        );
    }

    return recursiveObjectExpression(t, {
        queries: parsed.queries.map(query => ({
            query: removeQuery ? "" : query.query,
            queryName: query.queryName,
            queryId: query.queryId,
            usedFragments: query.usedFragments,
        })),
        fragments: parsed.fragments.map(fragment => ({
            fragment: removeQuery ? "" : fragment.fragment,
            fragmentName: fragment.fragmentName,
            fragmentId: fragment.fragmentId,
            usedFragments: fragment.usedFragments,
        })),
    });
}

const externalModules: {
    name: string;
    identifier?: string;
}[] = [
    {
        // import gql from 'graphql-tag'
        name: "graphql-tag",
    },
    {
        name: "graphql-tag.macro",
    },
    {
        // import { graphql } from 'gatsby'
        name: "gatsby",
        identifier: "graphql",
    },
    {
        name: "apollo-server-express",
        identifier: "gql",
    },
    {
        name: "apollo-server",
        identifier: "gql",
    },
    {
        name: "react-relay",
        identifier: "graphql",
    },
    {
        name: "apollo-boost",
        identifier: "gql",
    },
    {
        name: "apollo-server-koa",
        identifier: "gql",
    },
    {
        name: "apollo-server-hapi",
        identifier: "gql",
    },
    {
        name: "apollo-server-fastify",
        identifier: "gql",
    },
    {
        name: " apollo-server-lambda",
        identifier: "gql",
    },
    {
        name: "apollo-server-micro",
        identifier: "gql",
    },
    {
        name: "apollo-server-azure-functions",
        identifier: "gql",
    },
    {
        name: "apollo-server-cloud-functions",
        identifier: "gql",
    },
    {
        name: "apollo-server-cloudflare",
        identifier: "gql",
    },
];

function findIdentifier(importName: string) {
    return externalModules.find(m => m.name === importName);
}

export default function bemedBabelPlugin(
    babel: Babel,
): { visitor: Visitor<VisitorState> } {
    const t = babel.types;
    const isProduction = Boolean(process.env.NODE_ENV === "production");

    const qm = new QueryManager({
        async onExportQuery(query, target) {
            if (!target) {
                return;
            }

            await fs.mkdir(target, { recursive: true });

            await fs.writeFile(
                PathUtils.join(
                    target,
                    `${query.queryName}-${query.fullQueryId}.graphql`,
                ),
                query.fullQuery,
            );
        },
    });

    /**
     * Local name of the css import from babel-gql if any
     */
    let ourName: string | null = null;
    let externalTags = new Set<string>();
    let inject: any[] = [];

    return {
        visitor: {
            Program: {
                enter(path) {
                    // Reset import name state when entering a new file
                    ourName = null;
                    inject = [];
                    externalTags = new Set();
                },

                exit(path, state) {
                    if (!ourName) {
                        return;
                    }

                    for (const graphqObject of inject) {
                        path.pushContainer(
                            "body",
                            t.callExpression(t.identifier(ourName), [
                                graphqObject,
                            ]),
                        );
                    }
                },
            },

            ImportDeclaration(path, state) {
                const opts = state.opts || {};

                const target = opts?.ownModuleName ?? "babel-gql";
                const importName = path.node.source.value;

                if (importName === target) {
                    for (const s of path.node.specifiers) {
                        if (!t.isImportSpecifier(s)) {
                            continue;
                        }

                        if (s.imported.name === "gql") {
                            ourName = s.local.name;
                        }
                    }
                }

                const external = findIdentifier(importName);

                if (external) {
                    for (const s of path.node.specifiers) {
                        if (!t.isImportSpecifier(s)) {
                            continue;
                        }

                        if (
                            external.identifier &&
                            s.imported.name === external.identifier
                        ) {
                            externalTags.add(s.local.name);
                        }
                    }
                }
            },

            TaggedTemplateExpression(path, state) {
                if (!path.node.loc) {
                    return;
                }

                const tag = path.node.tag;

                if (!t.isIdentifier(tag)) {
                    return;
                }

                const removeQuery = state.opts?.removeQuery ?? isProduction;

                if (ourName && tag.name === ourName) {
                    const graphqObject = parseTag(t, qm, path, removeQuery);

                    path.replaceWith(
                        t.callExpression(t.identifier(ourName), [graphqObject]),
                    );
                } else if (externalTags.has(tag.name)) {
                    if (inject) {
                        inject.push(parseTag(t, qm, path, removeQuery));
                    }
                }

                const shouldExport = state.opts?.export ?? isProduction;

                if (shouldExport) {
                    qm.exportDirtyQueries(
                        state.opts?.target ??
                            PathUtils.join(process.cwd(), ".queries"),
                    );
                }
            },
        },
    };
}

function recursiveObjectExpression(t: typeof BabelTypes, ob: any): any {
    if (typeof ob === "string") {
        return t.stringLiteral(ob);
    }

    if (typeof ob === "boolean") {
        return t.booleanLiteral(ob);
    }

    if (ob === null) {
        return t.nullLiteral();
    }

    if (Array.isArray(ob)) {
        return t.arrayExpression(
            ob.map(item => recursiveObjectExpression(t, item)),
        );
    }

    if (typeof ob === "number") {
        return t.numericLiteral(ob);
    }

    if (!ob) {
        return t.nullLiteral();
    }

    return t.objectExpression(
        Object.keys(ob)
            .sort()
            .map(key => {
                return t.objectProperty(
                    t.identifier(key),
                    recursiveObjectExpression(t, ob[key]),
                );
            }),
    );
}
