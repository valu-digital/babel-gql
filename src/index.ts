import { ParsedGQLTag, combinedIds } from "./shared";

export { ParsedGQLTag };

interface TagListener {
    (tag: ParsedGQLTag): any;
}

export function createRuntimeGQL() {
    const queries: Record<string, ParsedGQLTag["queries"][0] | undefined> = {};
    const fragments: Record<
        string,
        ParsedGQLTag["fragments"][0] | undefined
    > = {};

    const listeners: TagListener[] = [];

    function registerGQLListener(fn: TagListener) {
        listeners.push(fn);
    }

    function gql(
        literals: TemplateStringsArray,
        ...placeholders: string[]
    ): ParsedGQLTag {
        if (!Array.isArray(literals)) {
            // Was converted to a normal function call via babel
            return runtimeGQL(literals as any);
        }

        throw new Error("babel-gql not configured?");
    }

    function runtimeGQL(parsed: ParsedGQLTag) {
        listeners.forEach((fn) => fn(parsed));

        parsed.queries.forEach((query) => {
            queries[query.queryName] = query;
        });
        parsed.fragments.forEach((fragment) => {
            fragments[fragment.fragmentName] = fragment;
        });

        return parsed;
    }

    function findFragmentsOfFragments(
        fragmentName: string,
        _frags: Record<string, boolean>,
    ): Record<string, boolean> {
        const fragment = fragments[fragmentName];

        if (!fragment) {
            throw new Error(`Cannot find fragment ${fragmentName}`);
        }

        if (!_frags) {
            _frags = {};
        }

        _frags[fragmentName] = true;

        fragment.usedFragments.forEach((fragmenName) => {
            _frags[fragmenName] = true;
            findFragmentsOfFragments(fragmenName, _frags);
        });

        return _frags;
    }

    return {
        gql,
        runtimeGQL,
        registerGQLListener,
        getQueries() {
            return queries;
        },
        getFragments() {
            return fragments;
        },
        getQuery(queryName: string) {
            const query = queries[queryName];

            if (!query) {
                throw new Error(`Cannot find query ${queryName}`);
            }

            const frags: Record<string, boolean> = {};

            query.usedFragments.forEach((fragmentName) => {
                findFragmentsOfFragments(fragmentName, frags);
            });

            const usedFragments = Object.keys(frags).map((fragmentName) => {
                const fragment = fragments[fragmentName];

                if (!fragment) {
                    throw new Error(
                        `Cannot find fragment ${fragmentName} for query ${queryName}`,
                    );
                }

                return fragment;
            });

            return {
                query: (
                    usedFragments.map((f) => f.fragment).join("\n") +
                    "\n" +
                    query.query
                ).trim(),
                queryName: queryName,
                queryId: combinedIds([
                    query.queryId,
                    ...usedFragments.map((f) => f.fragmentId),
                ]),
            };
        },
    };
}

function doRequest(
    endpoint: string,
    options: {
        query: {
            queryName: string;
            queryId: string;
            query: string;
        };
        variables?: object;
        fetch?: typeof fetch;
        fetchOptions?: RequestInit;
    },
) {
    const ourFetch = options.fetch ?? fetch;

    if (process.env.NODE_ENV !== "production") {
        const { headers, ...otherOptions } = options.fetchOptions ?? {};

        return ourFetch(endpoint, {
            method: "POST",
            headers: {
                "content-type": "application/json",
                ...options.fetchOptions?.headers,
            },
            ...otherOptions,
            body: JSON.stringify({
                variables: options.variables,
                operationName: options.query.queryName,
                query: options.query.query,
            }),
        });
    }

    const params = new URLSearchParams();
    params.append("operationName", options.query.queryName);
    if (options.variables) {
        params.append("variables", JSON.stringify(options.variables));
    }

    params.append(
        "extensions",
        JSON.stringify({
            persistedQuery: {
                version: 1,
                sha256Hash: options.query.queryId,
            },
        }),
    );

    // XXX turn to POST if mutation
    return ourFetch(endpoint + "?" + params.toString(), options.fetchOptions);
}

export function request<ResponseType = any>(
    endpoint: string,
    options: {
        query: string | ParsedGQLTag;
        variables?: object;
        fetch?: typeof fetch;
        fetchOptions?: RequestInit;
    },
): Promise<{
    response: Response;
    errors?: any[];
    data: ResponseType;
}> {
    const query = options.query;

    if (typeof query !== "string") {
        if (query.queries.length === 0) {
            throw new Error("Cannot find graphql query from tag");
        }

        if (query.queries.length > 1) {
            console.warn(
                "Multiple queries defined in request query " +
                    query.queries.map((q) => q.queryName).join(", "),
            );
        }
    }

    let promise;

    if (typeof query === "string") {
        promise = doRequest(endpoint, {
            query: getQuery(query),
            variables: options.variables,
            fetch: options.fetch,
            fetchOptions: options.fetchOptions,
        });
    } else {
        const queryOb = query.queries[0];
        promise = doRequest(endpoint, {
            query: getQuery(queryOb.queryName),
            variables: options.variables,
            fetch: options.fetch,
            fetchOptions: options.fetchOptions,
        });
    }

    let response: Response;
    return promise
        .then((res) => {
            response = res;
            return res.json();
        })
        .then((data) => {
            if (data.errors?.length ?? 0 > 0) {
                data.errors.forEach((error: any) => {
                    console.error(
                        "GraphQL response error:",
                        error.message,
                        error,
                    );
                });
            }
            data.response = response;
            return data;
        });
}

const babelqgl = createRuntimeGQL();

export const gql = babelqgl.gql;
export const getQuery = babelqgl.getQuery;
export const registerGQLListener = babelqgl.registerGQLListener;
export const getQueries = babelqgl.getQueries;
export const getFragments = babelqgl.getFragments;
