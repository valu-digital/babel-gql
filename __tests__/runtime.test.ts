import { createRuntimeGQL } from "../src";
import dedent from "dedent";

test("usage without babel", () => {
    const { gql } = createRuntimeGQL();

    const query = gql`
        query getFoo {
            foo
        }
    `;

    if (query.babel) {
        throw new Error("no babel here");
    }

    expect(dedent(query.query)).toEqual(dedent`
        query getFoo {
            foo
        }
    `);
});

test("runtime return value with babel", () => {
    const { runtimeGQL } = createRuntimeGQL();

    const query = runtimeGQL({
        queries: [{ queryId: "123", queryName: "getTest", usedFragments: [] }],
        fragments: [],
    });

    expect(query).toEqual({
        babel: true,
        queries: [{ queryId: "123", queryName: "getTest", usedFragments: [] }],
        fragments: [],
    });
});

test("can get query", () => {
    const { runtimeGQL, getQuery } = createRuntimeGQL();

    runtimeGQL({
        queries: [{ queryId: "123", queryName: "getTest", usedFragments: [] }],
        fragments: [],
    });

    expect(getQuery("getTest")).toEqual({
        queryId: "123",
        queryName: "getTest",
    });
});

test("can get query with fragments", () => {
    const { runtimeGQL, getQuery } = createRuntimeGQL();

    runtimeGQL({
        queries: [
            {
                queryId: "123",
                queryName: "getTest",
                usedFragments: ["NamedParts"],
            },
        ],
        fragments: [
            {
                fragmentId: "abc",
                fragmentName: "NamedParts",
                usedFragments: [],
            },
        ],
    });

    expect(getQuery("getTest")).toEqual({
        queryId: "1a",
        queryName: "getTest",
    });
});

test("can get query with fragments depending on fragments", () => {
    const { runtimeGQL, getQuery } = createRuntimeGQL();

    runtimeGQL({
        queries: [
            {
                queryId: "123",
                queryName: "getTest",
                usedFragments: ["NamedParts"],
            },
        ],
        fragments: [
            {
                fragmentId: "abc",
                fragmentName: "NamedParts",
                usedFragments: ["OtherFragment"],
            },
            {
                fragmentId: "efg",
                fragmentName: "OtherFragment",
                usedFragments: [],
            },
        ],
    });

    expect(getQuery("getTest")).toEqual({
        queryId: "1ae",
        queryName: "getTest",
    });
});

test("can use multiple runtime calls", () => {
    const { runtimeGQL, getQuery } = createRuntimeGQL();

    runtimeGQL({
        queries: [],
        fragments: [
            {
                fragmentId: "efg",
                fragmentName: "OtherFragment",
                usedFragments: [],
            },
        ],
    });

    runtimeGQL({
        queries: [
            {
                queryId: "123",
                queryName: "getTest",
                usedFragments: ["NamedParts"],
            },
        ],
        fragments: [
            {
                fragmentId: "abc",
                fragmentName: "NamedParts",
                usedFragments: ["OtherFragment"],
            },
        ],
    });

    expect(getQuery("getTest")).toEqual({
        queryId: "1ae",
        queryName: "getTest",
    });
});
