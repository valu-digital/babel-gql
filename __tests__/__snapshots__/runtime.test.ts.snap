// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can get query with fragments 1`] = `
"fragment NamedPars on Person { field2 }
query getTest { field, ...NamedParts }"
`;
